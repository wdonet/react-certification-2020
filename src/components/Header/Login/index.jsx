import React from 'react';
import wizelogo from '../../../img/wize-logo.png';

const Login = () => {
  const authenticated = true;

  return (
    <a className="login-link" href="_blank">
      <img src={wizelogo} alt="wizelogo" height={60} />
      {authenticated ? <div className="login-user">Mario</div> : <div>Sign in</div>}
    </a>
  );
};

export default Login;
