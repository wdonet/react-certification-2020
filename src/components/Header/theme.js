const base = {
  easeOutBack: 'cubic-bezier(0.34, 1.56, 0.64, 1)',
  colorWhite: 'rgb(255, 255, 255)',
  colorBlack: 'rgb(0, 0, 0)',
};

const dark = {
  id: 'dark',
  ...base,
  backgroundColor: '#333',
  textColor: 'black',
};

const light = {
  id: 'light',
  ...base,
  backgroundColor: '#666',
  textColor: 'white',
};

export const theme = { dark, light };
