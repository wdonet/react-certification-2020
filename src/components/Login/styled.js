import styled from 'styled-components';

const LoginBackground = styled.div`
  display: none;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0, 0, 0);
  background-color: rgba(0, 0, 0, 0.7);

  display: ${({ open }) => (open ? 'block' : 'none')};
`;

const LoginBox = styled.div`
  background-color: #fefefe;
  margin: 15% auto;
  padding: 20px;
  border: 1px solid #888;
  border-radius: 5%;
  width: 300px;
  height: 350px;
  position: relative;
`;

const CloseLoginButton = styled.span`
  color: #aaa;
  position: absolute;
  top: 0px;
  right: 15px;
  font-size: 28px;
  font-weight: bold;
  &:hover,
  &:focus {
    color: black;
    text-decoration: none;
    cursor: pointer;
  }
`;

const ErrorText = styled.p`
  color: red;
  font-size: 14px;
`;

const InputText = styled.input`
  width: 100%;
  height: 30px;
  border: 1px solid ${({ theme }) => theme.font.secondary};
  margin: 10px 0;
`;

const FormStyled = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 20px 0;
`;

export { LoginBackground, LoginBox, CloseLoginButton, ErrorText, InputText, FormStyled };
