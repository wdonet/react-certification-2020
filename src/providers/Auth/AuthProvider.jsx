import React, { useState, useEffect, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { loginUser } from '../../mocks/loginApi';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

function AuthProvider({ authenticatedJest, children }) {
  const [authenticated, setAuthenticated] = useState(authenticatedJest);

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setAuthenticated(isAuthenticated || authenticatedJest);
  }, []);

  const login = async ({ username, password }) => {
    try {
      const authorized = await loginUser({ username, password });
      if (authorized) {
        setAuthenticated(true);
        storage.set(AUTH_STORAGE_KEY, true);
      }
      return { success: true };
    } catch (error) {
      return { error };
    }
  };

  const logout = useCallback(() => {
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, false);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext };
export default AuthProvider;
