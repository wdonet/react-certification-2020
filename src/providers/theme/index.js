export { default, useTheme } from './Theme.provider';
