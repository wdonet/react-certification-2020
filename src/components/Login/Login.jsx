import React, { useContext, useState } from "react";
import styled from "styled-components";
import loginService from '../../services/loginService';
import AppContext from "../../providers/AppContext";
import TextField from "../../ui/TextField/TextField";
import Button from "../../ui/Button/Button";
import { useHistory } from "react-router";

const StyledLogin = styled.div`
    border-radius: 8px;
    padding: 4px;
    margin: 4px;
    height: min-content;
    width: 400px;
    background: ${ ({theme}) => theme.color.surface };
    color: ${ ({theme}) => theme.color.fontPrimary };
`

const StyledSection = styled.div`
    display: ${({display}) => display ? display : 'grid'};
    height: min-content;
    margin: 8px;
`;

const Login = (props) => {
    const { push } = useHistory();
    const { theme, setUserSession } = useContext(AppContext);
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [error, setError] = useState(null);
    const { onCancel } = props;

    const cleanUp = () => {
        document.getElementById("username").value = "";
        document.getElementById("password").value = "";
        setError("");
    }

    const loginAttempt = async () => {
        try{
            setError(null);
            const response = await loginService(username, password);
            cleanUp();
            setUserSession(response);
            push({ pathname: "/home" });
            onCancel && onCancel();
        }catch(error){
            cleanUp();
            setError(error.message);
        }
    }

    return (<StyledLogin role="form" theme={theme} data-testid="login-form">
                <StyledSection>Login!</StyledSection>
                <StyledSection>
                    <label htmlFor="username">User</label>
                    <TextField 
                        id="username" 
                        name="username"
                        data-testid="username-input"
                        onChange={(uname) => setUsername(uname)}
                    />
                </StyledSection>
                <StyledSection>
                    <label htmlFor="password">Password</label>
                    <TextField 
                        id="password" 
                        name="password"
                        data-testid="password-input"
                        type="password"
                        onChange={(pword) => setPassword(pword)}
                    />
                </StyledSection>
                <StyledSection>
                <span style={{color:'red'}}>{ error }</span>
                </StyledSection>
                <StyledSection display="flex">
                    <Button 
                        data-testid="login-button"
                        onClick={loginAttempt}
                        margin="4px"
                        height="30px"
                    >
                        Login
                    </Button>
                    <Button 
                        data-testid="cancel-button"
                        onClick={() => onCancel && onCancel()}
                        margin="4px"
                        height="30px"
                    >
                        Cancel
                    </Button>
                </StyledSection>
            </StyledLogin>);
}

export default Login;