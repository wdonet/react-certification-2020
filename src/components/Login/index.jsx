import React, { useEffect, useRef, useState } from 'react';

import {
  LoginBackground,
  LoginBox,
  CloseLoginButton,
  ErrorText,
  InputText,
  FormStyled,
} from './styled';

import { useAuth } from '../../providers/Auth';

const Login = ({ open, closeModal }) => {
  const [loginError, setLoginError] = useState(null);
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const { login } = useAuth();
  const loginBoxRef = useRef(null);
  useEffect(() => {
    const handleClickOutside = (e) => {
      if (loginBoxRef.current && !loginBoxRef.current.contains(e.target)) {
        closeModal();
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [loginBoxRef]);

  const loginUser = async () => {
    const { success } = await login({ username, password });
    if (!success) {
      setLoginError('Invalid username or password');
    } else {
      setLoginError(null);
      closeModal();
    }
  };

  const handleUsernameChange = (e) => {
    const {
      target: { value },
    } = e;
    setUsername(value);
  };

  const hanglePasswordChange = (e) => {
    const {
      target: { value },
    } = e;
    setPassword(value);
  };

  return (
    <LoginBackground open={open}>
      <LoginBox ref={loginBoxRef}>
        <CloseLoginButton onClick={closeModal}>&times;</CloseLoginButton>
        <FormStyled>
          <InputText
            name="username"
            value={username}
            placeholder="Username"
            onChange={handleUsernameChange}
          />
          <InputText
            name="password"
            placeholder="Password"
            type="password"
            value={password}
            onChange={hanglePasswordChange}
          />
          {loginError && (
            <ErrorText
              css={`
                color: ${({ theme }) => theme.palette.alert.main};
                padding: 2px;
                position: absolute;
                top: calc(100% - 25px);
                font-size: 16px;
              `}
            >
              {loginError}
            </ErrorText>
          )}
          <button
            type="button"
            onClick={() => {
              loginUser({ username: 'fulano', password: '1234567' });
              return false;
            }}
          >
            Login
          </button>
        </FormStyled>
      </LoginBox>
    </LoginBackground>
  );
};

export default Login;
