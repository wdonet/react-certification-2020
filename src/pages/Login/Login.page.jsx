import React, { useState } from 'react';

import { useAuth } from '../../providers/Auth';
import { Container, ErrorLabel } from './Login.page.styled';

function LoginPage() {
  const { login } = useAuth();
  const [values, setValues] = useState({});
  const [error, setError] = useState();

  const authenticate = async (event) => {
    event.preventDefault();
    try {
      setError(undefined);
      await login(values);
    } catch (err) {
      setError(err);
    }
  };

  const handleInputChange = (event) => {
    const newValues = { ...values };
    newValues[event.target.id] = event.target.value;
    setValues(newValues);
  };

  return (
    <Container>
      <h1>Welcome back!</h1>
      <form onSubmit={authenticate} className="login-form">
        <div className="form-group">
          <label htmlFor="username">
            <strong>username </strong>
            <input
              required
              type="text"
              id="username"
              value={values.username || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <div className="form-group">
          <label htmlFor="password">
            <strong>password </strong>
            <input
              required
              type="password"
              id="password"
              value={values.password || ''}
              onChange={handleInputChange}
            />
          </label>
        </div>
        <button type="submit">login</button>
      </form>
      {error && <ErrorLabel>Username or password incorrect!</ErrorLabel>}
    </Container>
  );
}

export default LoginPage;
