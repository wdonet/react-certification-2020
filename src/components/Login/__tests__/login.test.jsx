import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';

import Login from '../index';
import AuthProvider from '../../../providers/Auth';

import { lightTheme } from '../../../theme';

describe('Login', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <AuthProvider>
          <ThemeProvider theme={lightTheme}>
            <Login open closeModal={() => {}} />
          </ThemeProvider>
        </AuthProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
