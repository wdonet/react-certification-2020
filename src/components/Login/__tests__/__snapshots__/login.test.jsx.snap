// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
.c0 {
  display: none;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgb(0,0,0);
  background-color: rgba(0,0,0,0.7);
  display: block;
}

.c1 {
  background-color: #fefefe;
  margin: 15% auto;
  padding: 20px;
  border: 1px solid #888;
  border-radius: 5%;
  width: 300px;
  height: 350px;
  position: relative;
}

.c2 {
  color: #aaa;
  position: absolute;
  top: 0px;
  right: 15px;
  font-size: 28px;
  font-weight: bold;
}

.c2:hover,
.c2:focus {
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c4 {
  width: 100%;
  height: 30px;
  border: 1px solid #666666;
  margin: 10px 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 20px 0;
}

<div
  className="c0"
  open={true}
>
  <div
    className="c1"
  >
    <span
      className="c2"
      onClick={[Function]}
    >
      Ã—
    </span>
    <form
      className="c3"
    >
      <input
        className="c4"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        value=""
      />
      <input
        className="c4"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        onClick={[Function]}
        type="button"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
